### ISSUE TITLE
Crash when building or trying to open .gui files on Ubuntu 20.04 & 20.10

### ISSUE THREAD

**Issue by lnmp4000** [defold/defold#5342]
### Expected behaviour
Should allow editing of .gui files and should build and launch the project.

### Actual behaviour

Crash with:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fc54149d604, pid=48082, tid=48932
#
# JRE version: OpenJDK Runtime Environment (11.0.1+13) (build 11.0.1+13)
# Java VM: OpenJDK 64-Bit Server VM (11.0.1+13, mixed mode, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# C  [jna394967780279324120.tmp+0x7604]
```
[hs_err_pid48082.log](https://github.com/defold/defold/files/5548445/hs_err_pid48082.log)

### Steps to reproduce

Open editor and load project. Find a .gui file and double click. The Defold editor then crashes.

<hr/>

<table>
<tr><td>Build time</td><td>2020-11-15T13:58:09.322179</td></tr>
<tr><td>Defold channel</td><td>editor-alpha</td></tr>
<tr><td>Defold editor sha</td><td>eb31ede76a85f9e7dd669c001f4b5d0aa4e2787a</td></tr>
<tr><td>Defold engine sha</td><td>e41438cca6cc1550d4a0131b8fc3858c2a4097f1</td></tr>
<tr><td>Defold version</td><td>1.2.175</td></tr>
<tr><td>GPU</td><td>GeForce 930MX/PCIe/SSE2</td></tr>
<tr><td>GPU Driver</td><td>4.6.0 NVIDIA 455.28</td></tr>
<tr><td>Java version</td><td>11.0.1+13</td></tr>
<tr><td>OS arch</td><td>amd64</td></tr>
<tr><td>OS name</td><td>Linux</td></tr>
<tr><td>OS version</td><td>5.8.0-28-generic</td></tr>
<table>

**Comment by Cryoge** [comment]
I seem to have trouble with this as well, i have tried multiple older versions of the defold editor and all throw the same error.

```
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f253e89b604, pid=30799[thread 30850 also had an error], tid=30854

#
# JRE version: OpenJDK Runtime Environment (11.0.1+13) (build 11.0.1+13)
# Java VM: OpenJDK 64-Bit Server VM (11.0.1+13, mixed mode, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# C
  [jna13566827014821332839.tmp+0x7604]
#
```

Log:
[hs_err_pid30799.log](https://github.com/defold/defold/files/5556578/hs_err_pid30799.log)

**Comment by tewald** [comment]
Happens on Ubuntu 20.10 too.
Seems related do libffi version.
[hs_err_pid4093.log](https://github.com/defold/defold/files/5556667/hs_err_pid4093.log)

**Comment by thejustinwalsh** [comment]
Maybe an approach like this will work?
https://github.com/cryptomator/cryptomator-linux/issues/27

**Comment by Cryoge** [comment]
> Maybe an approach like this will work?
> [cryptomator/cryptomator-linux#27](https://github.com/cryptomator/cryptomator-linux/issues/27)

Yes!!! This worked for me! Thank you very much for pointing me in the right direction!

Running Defold with the LD_PRELOAD environment variable pointing to a older version of libffi (like version 6 or 7) i managed to fix the issue. (Atleast for Pop!_os 20.10 but i assume it will also work on Ubuntu or any other Debian based system)

We need to have libffi6 or libffi7 installed on our system. You can check under `/usr/lib/x86_64-linux-gnu` if you can find `libffi.so.6` or `libffi.so.7`

If none of those versions are available then you can install version 7:
`
wget http://ftp.br.debian.org/debian/pool/main/libf/libffi/libffi7_3.3-5_amd64.deb
sudo dpkg -i libffi7_3.3-5_amd64.deb
`

Now you can run Defold like this: (make sure to change the path of LD_PRELOAD if using version 6)
`
LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libffi.so.7 ./Defold
`

**Comment by lnmp4000** [comment]
I can confirm this works on Ubuntu 20.10.
I had a choice of `libffi.so.8` and  `libffi.so.6` `.8` does not work, `.6` does.

So setting the `LD_PRELOAD` to
`LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libffi.so.6`
is all you need.

**Comment by britzl** [comment]
Thank you for confirming this works. But surely there must be some way to solve this without these additional steps?

I guess we could include a defold.sh or something that does this based on what is available etc but is that really the best solution?

**Comment by tewald** [comment]
Using version 7 manualy downloaded and installed it works fine.
With default version 8 used on Ubuntu 20.10, doesn't.
It's a workaround, not very nice, but works for now.

**Comment by thejustinwalsh** [comment]
Still a workaround, but you can also create a `.desktop` file with this command in it and it works as well.

```.desktop
[Desktop Entry]
Name=Defold
Comment=Defold Game Engine
Exec=env LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libffi.so.7 /home/tjw/Applications/Defold/Defold %U
Terminal=false
Type=Application
Icon=/home/tjw/Applications/Defold/defold.png
StartupWMClass=Defold
Categories=Development;
TryExec=/home/tjw/Applications/Defold/Defold
PrefersNonDefaultGPU=true
```
I already do this for launching on my discrete GPU on the laptop, so easy workaround for me.

---

As to the issue at hand, it sounds like a java dependency was built against or bundles an included version of libffi and depending on execution order the system version of libffi is loaded before the bundled version, and you get into a situation where you are executing calls through libffi v8 when it should be v6 / v7?

I am not sure if this is an issue with the JDK, libffi v8, or some expectation of load order that is being violated in the code itself. I couldn't determine what was loading libffi v8 vs v7 and why the linked workaround issue needed to extract libffi from the jar in the first place.

**Comment by cnkuyan** [comment]
I've been getting the same crash on ubuntu 20.10, very similar coredump - but in my case referencing libopenal . 
Pointing the library to an older version is a non-fix , IMHO .  

the issue is still there, and effectively blocks devs from advancing as long as the 3rd party dependency exists.

Thanks

**Comment by britzl** [comment]
> I've been getting the same crash on ubuntu 20.10, very similar coredump - but in my case referencing libopenal .

Is this when running a built game or when working with the editor?

**Comment by cnkuyan** [comment]
when working with the editor.

I believe you are the author of the examples under the repo : 
https://github.com/britzl/publicexamples/tree/master/examples

On all of the example projects , it crashed on me after I double clicked on any .collection file .  After showing the initial display for a while, it crashes with the coredump.  Again , the OS : ubuntu 20.10

Thanks
I can run the

**Comment by britzl** [comment]
> when working with the editor.
> in my case referencing libopenal

Can you please share the full crash details?

**Comment by britzl** [comment]
Closing this as we no longer support Ubuntu 20. Lowest version we support is 22.04 LTS.
